### ISSUE TITLE
How to stop receiving data streams? 

### ISSUE THREAD

**Issue by MaggieWang1011** [sccn/labstreaminglayer#198]
Hi,
I am a beginner on this and hope someone can help me here. The question is how to stop receiving data streams?
In me project I have three data streams (EEG, audio, mouse) in the network, and I only want to record data while the subject is performing the required task. There are some time intervals between task trials for resting that I don't want data in these time intervals. 
Is there a way to stop receiving data streams? Or to control the received data streams?

Thank you very much.

Maggie

**Comment by cboulay** [comment]
Hi Maggie,

LSL doesn't allow inlets to control the originating stream. You have some options to achieve what you are suggesting.

* Outlets can be configured to transmit at irregular_rate, and the outlet can choose to only push samples when it thinks it is relevant to do so.
* The application with the inlets can pull_chunk but then simply ignore what it pulls when it doesn't need to do anything.
* The Python version of LabRecorder has an interface that allows it to be controlled programmatically so data files can be closed and opened automatically. However, this version of LabRecorder is deprecated and was relatively unstable.

What are you trying to achieve by stopping the data streams? Are your data files getting too large?

**Comment by MaggieWang1011** [comment]
Thank you for replying, Chadwick.

I have an experiment to record EEG, audio and video when the subjects are performing a task. I want to write data to a .txt file (or .xdf I wish, but I don't know how) second by second, stop writing at the resting interval, then start writing again at next task trial. To be more exact, I am not trying to stop the data stream, but to have some control over the streams. Can I do that?

And is there a C# version of LabRecorder that allows data files to be controlled automatically.

I am keen on C# becaese I developed my experiment software using C# and I have to synchronize the task with data recording. For example, when the subject click 'Button1', a task trial begins and data recording  (EEG, video, audio) should start simultaneously; when this trial ends, data recording should also stop.

Do you have any suggestions?

Maggie

**Comment by cboulay** [comment]
The data can be recorded to xdf by using the LabRecorder application that comes with LSL. I've never tried the scripting interface so I don't know what would be involved in controlling the Python version of LabRecorder; I only use the C++ version that currently does not have a mechanism to auto-start or auto-stop.

I don't know what kind of control you want over a stream, other than stopping it. Can you elaborate?

In practice, it's a good idea to store data between trials even if you aren't planning on analyzing it. It can be useful for offline analysis as baseline data and this can have many uses. The only reason I'd ever suggest not storing data between trials is if your data files are becoming too big to open and your downtime is really long. But in this case I would close the data file after N trials then open a new one for the next trials. Hard drives are cheap relative to the expense in collecting good data.

If you plan on re-analyzing your data many times, then the first step in your offline analysis can be to make an intermediate data file that only contains minimal data of interest (it might be a good idea to filter and downsample here too). That way you can have a much smaller file to work off of but you'll always have the raw data in case you ever need to go back to it.

**Comment by MaggieWang1011** [comment]
Thanks a lot. Good suggestions. 

But the problem is that if I keep storing data all the time, how could I know which part belongs to which trial or interval? Is there a way to make a marker or something like that?

By saying control, I mean simple things like when to start/stop writing data to a .txt file.

**Comment by cboulay** [comment]
Oh yes, your data file absolutely should have markers in it to tell you when events are happening. Any event that is perceptible to the participant should be marked, and anything the participant does should be marked if it is not already captured by some behavioural measurement. In LSL environments, there is usually a Markers stream that contains strings whenever something happens.

What software are you using to present task stimuli?

**Comment by MaggieWang1011** [comment]
I developed a WinForm application to present task stimuli using visual studio 2013 by C#.

**Comment by cboulay** [comment]
OK then you should make your application LSL-enabled and then send strings via a Markers outlet whenever an event happens, and make sure you store this stream in your datafile.

Please use [LabRecorder](ftp://sccn.ucsd.edu/pub/software/LSL/Apps/LabRecorder-1.12c.zip) and don't save to a txt file. (For future readers, the LabRecorder link might be outdated by the time you read this).

**Comment by MaggieWang1011** [comment]
The whole duration of my experiemnt may be 1.5 hour, which will end up with a really huge file. If I'd like to cut it into several files like you suggested, do I have other choice than close the current file and open a new one manually? Since the LabRecorder does not seem like to support open a new file other than manually.

**Comment by cboulay** [comment]
Well, you could take the source code from LabRecorder and customize it so that it scans the Markers stream looking for certain commands (e.g.: "LabRecorder:closeFile") and then acts accordingly. You would then modify your experiment to send those command on the Markers stream. This might be more work than its worth just for your personal use.

Take a read at issue #120 for some other suggestions. Post there to task @dmedine how the Lua interface would work (I'm curious too). You're not the only one that would like this feature.

**Comment by MaggieWang1011** [comment]
Thanks a lot for your information. Really helps.

**Comment by dmedine** [comment]
A couple of comments:

1. This:

|while (true) { int num1 = inlet_1.pull_chunk(buffer1, timestamps1); int 
num2 = inlet_2.pull_chunk(buffer2, timestamps2); int num3 = 
inlet_3.pull_chunk(buffer3, timestamps3); } |||

||is bad. Each of those streams has different sampling rates and every 
time you call to pull, it blocks until it is done. Thus it could be the 
case that you are blocking for a very long time (in the case of your 
mouse event stream) until you pull data from your EEG stream. The 
default buffer size for an outlet is 6 minutes, so as long as you don't 
block for longer than 6 minutes, you won't lose data, but you might. 
Also, when the outlet buffer fills up, it has to keep allocating memory. 
This is not an efficient way to proceed.

The way labrecorder gets around this problem is that when it finds a new 
stream to connect to, it puts each inlet object on it's own thread. This 
way calls to pull happen in parallel rather than in order.

2. There is no Lua implementation in LabRecorder. That is on our 
wishlist. You could try the older python implementation of labrecorder: 
ftp://sccn.ucsd.edu/pub/software/LSL/Apps/LabRecorder-1.10.zip which 
(obviously) supports python scripting. There is an example of this in 
the config file.

3. Chadwick is definitely right about putting LSL marker streams in your 
presentation software and emitting a marker every time you emit an 
event. He is also right about recording data at down time (unless it is 
excessively long). Don't worry about 1.5 hour recordings. This is 
commonplace. It will take a long time to load the data initially, but 
that (and storage expense) is the only penalty you will incur.

4. Implementing the xdf specification is not terribly hard, but it does 
require careful attention to detail. In general, I don't see any reason 
to reinvent the wheel here. I get the sense that all the tools you need 
are already there, you just need to get acquainted with them. Probably 
this is easier than re-writing LabRecorder in C#---although, since the 
core of LabRecorder is a C++ dll (RecorderLib), you could wrap it as 
either a COM object or some such other C# friendly interface. Then you 
could have C# control over the recording procedure without having to 
re-implement the connection process, the launching of threads, and the 
data formatting.

Hope this helps some!

**Comment by MaggieWang1011** [comment]
Thank you very much for these helpful information, David.

As a feedback. I made following changes to my experiment software, then everything is working fine. Now all the data streams are scychronized.

1) It creates a seperate thread for each new data stream it finds, so each inlet object is on its own thread which means data is being read in parallel.

2) A marker stream is developed to mark events.

3) Besides using LabRecorder for recording, I also recorded some segments of interest using my experiment software itself. Data streams were being read and written into seperate files within their own threads. Data streams were being read all the time and I only controlled when to start/stop writting. 

After two days' confusion, I am getting acquainted with LSL.

Thanks again.

Maggie
